src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,293 @@
+    }
+
+    trainGenerator = true;
+    // Training Generator
+    Generate(batchSize, std::move(fakeData), std::move(noiseData));
+    fakeLables.zeros(1, batchSize);
+    this->predictors = std::move(fakeData);
+    this->responses = std::move(fakeLables);
+    numFunctions = predictors.n_cols;
+    discriminator.predictors = predictors;
+    discriminator.responses = responses;
+    Optimizer.MaxIterations() = numFunctions;
+    Optimizer.Optimize(*this, parameter);
+
+    // set train generator to false
+    trainGenerator = false;
+  }
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+double GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Evaluate(const arma::mat& /*parameters*/,
+    const size_t i, const bool /*deterministic*/)
+{
+  currentInput = discriminator.predictors.unsafe_col(i);
+  currentTarget = discriminator.responses.unsafe_col(i);
+  discriminator.Forward(std::move(currentInput));
+  double res = discriminator.outputLayer.Forward(std::move(boost::apply_visitor(
+      outputParameterVisitor, discriminator.network.back())),
+      std::move(currentTarget));
+  return res;
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Gradient(const arma::mat& parameters, const size_t i, arma::mat& gradient)
+{
+  if (gradient.is_empty())
+  {
+    if (parameter.is_empty())
+    {
+      Reset();
+    }
+    gradient = arma::zeros<arma::mat>(generator.Parameters().n_rows +
+        discriminator.Parameters().n_rows, 1);
+  }
+  else
+  {
+    gradient.zeros();
+  }
+
+  // Gradient for generator network
+  gradientGenerator = arma::mat(gradient.memptr(),
+      generator.Parameters().n_rows,
+      generator.Parameters().n_cols, false, false);
+
+  // Gradient for discriminator network
+  gradientDisriminator = arma::mat(gradient.memptr(),
+      discriminator.Parameters().n_rows,
+      discriminator.Parameters().n_cols, false, false);
+
+  // get the gradients of the discriminator
+  discriminator.Gradient(parameters, i, gradientDisriminator);
+  // apply error to the first layer
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, discriminator.network.front())), std::move(
+      boost::apply_visitor(deltaVisitor, discriminator.network[1])), std::move(
+      boost::apply_visitor(deltaVisitor, discriminator.network.front()))),
+      discriminator.network.front());
+
+  if (trainGenerator)
+  {
+    // pass the error from discriminator to generator
+    generator.error = boost::apply_visitor(deltaVisitor,
+        discriminator.network[0]);
+    generator.Backward();
+    generator.ResetGradients(gradient);
+    generator.currentInput = noiseData.col(i);
+    generator.Gradient();
+  }
+
+  // Freeze weights of discrminator if generator is training
+  if (!trainGenerator)
+    gradientGenerator.zeros();
```

Comment:
The gradient is zero for each network, since you shared the gradient, take a look at the comment above.